File: mobile/lib/models/pantry_item.dart
import 'package:isar/isar.dart';

part 'pantry_item.g.dart';

/// Storage location for a pantry item.
enum StorageLocation {
  pantry,
  fridge,
  freezer,
}

@collection
class PantryItem {
  Id id = Isar.autoIncrement;

  late String name;

  late DateTime expiryDate;

  @enumerated
  late StorageLocation storageLocation;

  String? category;

  String? barcode;

  late DateTime createdAt;

  late DateTime updatedAt;

  @Index()
  late bool synced;

  // Constructor
  PantryItem({
    required this.name,
    required this.expiryDate,
    this.storageLocation = StorageLocation.fridge,
    this.category,
    this.barcode,
  }) {
    createdAt = DateTime.now();
    updatedAt = DateTime.now();
    synced = false;
  }

  // Empty constructor for Isar
  PantryItem.empty();

  /// Returns the number of days until this item expires.
  int get daysUntilExpiry =>
      expiryDate.difference(DateTime.now()).inDays;

  /// Returns true if the item is expired.
  bool get isExpired => daysUntilExpiry < 0;

  /// Returns true if the item expires within [days] days.
  bool isExpiringSoon(int days) => daysUntilExpiry <= days && !isExpired;

  /// Converts to JSON for API sync.
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'expiry_date': expiryDate.toIso8601String(),
      'storage_location': storageLocation.name,
      'category': category,
      'barcode': barcode,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }

  /// Creates a PantryItem from JSON (API response).
  factory PantryItem.fromJson(Map<String, dynamic> json) {
    final item = PantryItem(
      name: json['name'] as String,
      expiryDate: DateTime.parse(json['expiry_date'] as String),
      storageLocation: StorageLocation.values.firstWhere(
        (e) => e.name == json['storage_location'],
        orElse: () => StorageLocation.fridge,
      ),
      category: json['category'] as String?,
      barcode: json['barcode'] as String?,
    );
    if (json['id'] != null) item.id = json['id'] as int;
    if (json['created_at'] != null) {
      item.createdAt = DateTime.parse(json['created_at'] as String);
    }
    if (json['updated_at'] != null) {
      item.updatedAt = DateTime.parse(json['updated_at'] as String);
    }
    item.synced = true;
    return item;
  }
}

File: server/internal/models/pantry_item.go
package models

import (
	"time"

	"gorm.io/gorm"
)

// PantryItem represents a food item in the user's pantry.
type PantryItem struct {
	ID              uint           `gorm:"primaryKey" json:"id"`
	UserID          uint           `gorm:"index;not null" json:"user_id"`
	Name            string         `gorm:"size:255;not null" json:"name"`
	ExpiryDate      time.Time      `gorm:"index;not null" json:"expiry_date"`
	StorageLocation string         `gorm:"size:50;not null" json:"storage_location"` // pantry, fridge, freezer
	Category        string         `gorm:"size:100" json:"category,omitempty"`
	Barcode         string         `gorm:"size:50" json:"barcode,omitempty"`
	CreatedAt       time.Time      `json:"created_at"`
	UpdatedAt       time.Time      `json:"updated_at"`
	DeletedAt       gorm.DeletedAt `gorm:"index" json:"-"`
}

// WasteLog records a discarded food item for waste tracking.
type WasteLog struct {
	ID             uint      `gorm:"primaryKey" json:"id"`
	UserID         uint      `gorm:"index;not null" json:"user_id"`
	ItemName       string    `gorm:"size:255;not null" json:"item_name"`
	Category       string    `gorm:"size:100" json:"category,omitempty"`
	EstimatedPrice float64   `gorm:"type:decimal(10,2);not null" json:"estimated_price"`
	WastedDate     time.Time `gorm:"index;not null" json:"wasted_date"`
	CreatedAt      time.Time `json:"created_at"`
}
